File: others/compstr/compress_string.py
class CompressString(object):
    def compress(self, str):
        if str is None or not str:
            return str

        compstr, prev, count = '', str[0], 1

        for c in str:
            if c == prev:
                count += 1
            else:
                compstr += self._calc_partial_compstr(prev, count)
                prev = c
                count = 1

        compstr += self._calc_partial_compstr(prev, count)
        return compstr if len(compstr) < len(str) else str

    def _calc_partial_compstr(self, prev, count):
        return prev + (str(count) if count > 1 else '')

File: others/compstr/compress_string.go
package compstr

import "fmt"

type CompressString struct{}

func NewCompressString() CompressString {
	return CompressString{}
}

func (comp *CompressString) Compress(str string) string {
	n := len(str)
	if n < 1 {
		return str
	}

	compStr, prev, count := "", rune(str[0]), 0

	for _, c := range str {
		if c == prev {
			count++
		} else {
			compStr += comp.calcPartialCompStr(prev, count)
			prev = c
			count = 1
		}
	}

	compStr += comp.calcPartialCompStr(prev, count)

	fmt.Println(compStr)

	if len(compStr) < len(str) {
		return compStr
	}

	return str
}

func (comp *CompressString) calcPartialCompStr(prev rune, count int) string {
	if count > 1 {
		return fmt.Sprint(string(prev), count)
	}

	return fmt.Sprint(string(prev))
}
